### PR TITLE
[Clang] Fix handling of placeholder variables name in init captures

### LINKED ISSUE #107024: [clang] "error: ambiguous reference to placeholder" when '_' is used as identifier in lambda capture
### Minimized reproducer

```c++
int foo() {
    return [_ = 42]() { return _; }();
}
```
https://godbolt.org/z/5r7GW53re

### Error

```
<source>:2:32: error: ambiguous reference to placeholder '_', which is defined multiple times
    2 |     return [_ = 42]() { return _; }();
      |                                ^
<source>:2:13: note: placeholder declared here
    2 |     return [_ = 42]() { return _; }();
      |             ^
<source>:2:13: note: placeholder declared here
<source>:2:12: error: cannot initialize return object of type 'int' with an rvalue of type 'void'
    2 |     return [_ = 42]() { return _; }();
      |            ^~~~~~~~~~~~~~~~~~~~~~~~~~
2 errors generated.
```

### Further info

Affected versions: 18.1.0, trunk

clang 17.0.1 works ok, other compilers also seem to work ok

### PR DESCRIPTION
**Pull request by cor3ntin**
We were incorrectly not deduplicating results when looking up `_` which, for a lambda init capture, would result in an ambiguous lookup.

The same bug caused some diagnostic notes to be emitted twice.

Fixes #107024

### CODE REVIEW

**Review by zyn0217 — APPROVED** [review]
One thing I'm not entirely clear about is why we would end up with two same placeholder `Decl`s in `LookupResult`.

It seems that `IdResolver` results in two same duplicate Decls for such case, but I don't know why.

(However I'd still approve it because this reverts the condition to the state before the placeholder patch.)

**Comment by zyn0217** [comment]
I took a cursory look, and it seems that we pushed the same init capture to the scope chain three times:

1. `Sema::ActOnLambdaExpressionAfterIntroducer`
https://github.com/llvm/llvm-project/blob/6c59dfb0184aa11ec202f1cd8aee9e971a1565a4/clang/lib/Sema/SemaLambda.cpp#L1212-L1214

2. `Sema::ActOnLambdaExpressionAfterIntroducer`

https://github.com/llvm/llvm-project/blob/6c59dfb0184aa11ec202f1cd8aee9e971a1565a4/clang/lib/Sema/SemaLambda.cpp#L1324-L1327

3. `Sema::ActOnStartOfLambdaDefinition`
https://github.com/llvm/llvm-project/blob/6c59dfb0184aa11ec202f1cd8aee9e971a1565a4/clang/lib/Sema/SemaLambda.cpp#L1468-L1475

I'm confused if they're all necessary...

File: clang/lib/Sema/SemaLambda.cpp [hunk 1]
```diff
@@ -1323,7 +1323,6 @@ void Sema::ActOnLambdaExpressionAfterIntroducer(LambdaIntroducer &Intro,
 
     if (C->Init.isUsable()) {
       addInitCapture(LSI, cast<VarDecl>(Var), C->Kind == LCK_ByRef);
-      PushOnScopeChains(Var, CurScope, false);
```
Comment (zyn0217) [inline clang/lib/Sema/SemaLambda.cpp:1326]: Yeah, 2 is a duplicate of 1, but 3 is necessary because that constitutes a lambda body scope.

**Review by zyn0217 — APPROVED** [review]

**Comment by cor3ntin** [comment]
The second was clearly duplicated.
The first one introduce a scope for the captures, so that captures are available in the declaration of the lambda (before its compound statement).

But the body introduces a new scope wherein we inject the parameters - and we also need to reinject the captures to make sure they cannot be shadowed by a local variables

````cpp
[a = 0] 
// capture scope
{
 // body scope
 int a = 42; // error!
}
 ```

**Comment by cor3ntin** [comment]
Should we backport this?

**Comment by zyn0217** [comment]
I think so, yes.

**Comment by nikic** [comment]
For reference, the manual cherry pick is https://github.com/llvm/llvm-project/pull/107214.

**Pull request merged**